app_check/src/desktop/debug_provider_desktop.cc

```diff
@@ -14,19 +14,118 @@
 
 #include "app_check/src/desktop/debug_provider_desktop.h"
 
+#include <map>
+#include <string>
+#include <utility>
+
+#include "app/rest/response.h"
+#include "app/rest/transport_builder.h"
+#include "app/rest/transport_curl.h"
+#include "app/rest/util.h"
+#include "app/src/log.h"
+#include "app/src/scheduler.h"
+#include "app_check/src/desktop/debug_token_request.h"
+#include "app_check/src/desktop/token_response.h"
 #include "firebase/app_check/debug_provider.h"
 
 namespace firebase {
 namespace app_check {
 namespace internal {
 
-DebugAppCheckProviderFactoryInternal::DebugAppCheckProviderFactoryInternal() {}
+class DebugAppCheckProvider : public AppCheckProvider {
+ public:
+  explicit DebugAppCheckProvider(App* app);
+  ~DebugAppCheckProvider() override;
+
+  void GetToken(std::function<void(AppCheckToken, int, const std::string&)>
+                    completion_callback) override;
+
+ private:
+  App* app_;
+
+  scheduler::Scheduler scheduler_;
+};
+
+DebugAppCheckProvider::DebugAppCheckProvider(App* app)
+    : app_(app), scheduler_() {
+  firebase::rest::util::Initialize();
+  firebase::rest::InitTransportCurl();
+}
+
+DebugAppCheckProvider::~DebugAppCheckProvider() {
+  firebase::rest::CleanupTransportCurl();
+  firebase::rest::util::Terminate();
+}
+
+// Performs the given rest request, and calls the callback based on the
+// response.
+void GetTokenAsync(SharedPtr<DebugTokenRequest> request,
+                   std::function<void(AppCheckToken, int, const std::string&)>
+                       completion_callback) {
+  TokenResponse response;
+  firebase::rest::CreateTransport()->Perform(*request, &response);
 
-DebugAppCheckProviderFactoryInternal::~DebugAppCheckProviderFactoryInternal() {}
+  if (response.status() == firebase::rest::util::HttpSuccess) {
+    // Call the callback with the response token
+    AppCheckToken token;
+    token.token = std::move(response.token());
+    // Expected response is in seconds
+    int64_t extra_time = strtol(response.ttl().c_str(), nullptr, 10);
+    token.expire_time_millis = (response.fetch_time() + extra_time) * 1000;
+    completion_callback(token, kAppCheckErrorNone, "");
+  } else {
+    // Create an error message, and pass it along instead.
+    AppCheckToken token;
+    char error_message[1000];
```

Comment:
It seems a bit strange to me that we are working with a simple character array here when the callback wants a string, but this does seem to truncate the response body to at most 1000 characters which is nice. 
(No action required)

Reply:
Yeah, seemed like the safest way to handle it.